text/0020-contributing-guidelines.md

```diff
@@ -0,0 +1,46 @@
+
+[Github squash and merge](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-request-merges#squash-and-merge-your-pull-request-commits) is enabled for all frontend projects.
+
+For frontend projects within OpenMRS, the following guidelines apply:
+- Not all PRs have to link to a JIRA ticket.
+- Pull requests are allowed (and encouraged) to contain more than one commit. Those commits are squashed when merged into the master branch.
+- Any pull request template within the repo should be followed.
+
+## Definition
+In this PR, contribution refers to people creating pull requests to Github repos. Other terms in the definition section have been linked to above.
+
+## Reason for decision
+- Github contributing.md and pull_request_template.md markdown files are the worldwide standard for open source software. They may link to a separate guide, but should be present.
+- The OpenMRS code of conduct helps provide a safe environment for people within the OpenMRS community.
+- The OpenMRS MPL-2.0 + Health care disclaimer software license protects the freedom of the open source software while also protecting OpenMRS from lawsuits.
+- The OpenMRS pull request tips are a community maintained set of good practices.
+- Github squash and merge creates the most easily read and understood commit history on the master branch.
+- Enforcing that every PR links to a JIRA ticket discourages contribution by increasing barrier to entry. First, you have to create a JIRA account. Then you have to create an issue in the right project. Then you have to self assign it. Then you have to link to it. Then you have to close it. Linking to JIRA is appropriate and helpful when you are working on a ticket. However, not every code change is described by a JIRA ticket. See #19 for more details on the role of JIRA.
+- Enforcing that every PR has only one commit increases the barrier to entry while destroying git history. It makes it harder for the reviewer to see incremental progress in a PR. Due to Github's squash and merge feature, it is not required for the master branch in order to have a clean commit history.
+
+## Alternatives
+The OpenMRS community up already adopts much of what is in this PR. The new aspects of it are using Github squash and merge, not enforcing that all PRs are associated with a JIRA ticket, and allowing for multiple commits in a single PR. These new aspects only apply to frontend code.
+
+## Common practices (not enforced)
+- Do not increase the scope of a PR after it has been reviewed.
+- Encourage feedback as both an author and reviewer.
+- Do not close a PR and recreate a new one with the same code.
+- Encourage as many contributors to engage in deep and meaningful contribution.
+- Be sensitive to the varying experience levels of OpenMRS contributors.
+- Avoid gatekeeping, which includes over-emphasizing to contributors how they're failing to follow the correct process.
```

Comment:
Do you mind giving an example of how one can over-emphasise to contributors how they're failing to follow the correct process? Again, my question is to simply understand the above. 😊

Reply:
It's about trying to encourage first-time contributors. Here's a situation where I think overemphasizing process can be damaging and discourage contribution:

1. Software developer who is new to OpenMRS gets excited about finding a bug and wanting to fix it.
2. They submit a pull request, and even get the linter and tests and typescript checks to work.
3. Two or more maintainers of a project comment only on how they failed to link to a JIRA ticket, how they should have only had one commit in the PR, how they didn't follow the PR template, and how they need those things before the reviewer will even review the code.
4. Author feels a bit discouraged because they worked hard on the PR and all they heard was how they did it wrong. And the feedback wasn't even about what they were trying to fix. They go ahead and update the PR.
5. Reviewer now provides feedback and asks the author to make major changes to their PR.
6. It's now a week later and the contributor has lost interest / enthusiasm.

Reply:
Oh i see!
The challenge here is that we are trying to encourage code reviews from as many people as possible, including new comers who may not yet fully understand the code base, but can catch some of the basic things like those required by the pull request template. We do not want to discourage them from doing that kind of contribution. We usually give them a chance to do exactly such, before we get in with more detailed reviews. If the pull request author responds to these, but waits a full week or more, without detailed reviews, it is a result of not having enough reviewers, as per our current pull request back log which is growing every day: https://github.com/pulls?page=1&q=is%3Aopen+is%3Apr+user%3Aopenmrs&user=openmrs

Reply:
I think even when a maintainer even just says "Thank you for your contribution" inside of the main PR review comment that that can go a long way towards keeping the contributor excited.

It's very true it's hard to be a good reviewer when short on resources - I respond to about 20-30 github notifications every day and definitely get tired out by it. I empathize with and respect what you guys are doing on the java side with all your code reviews. 👍